File: wrapper/cctools_wrapper.py
import os
import sys
import subprocess
import shlex
from sys import exit

DIR=os.path.dirname(__file__)

def win_path_to_wsl_path(win_path):
    if sys.platform not in ('win32', 'cygwin'):
        return win_path
    win_path = win_path.replace('\\', '/')
    if not os.path.isabs(win_path):
        return win_path
    if win_path.startswith('/'):
        # MSYS2/Cygwin path, convert it to Windows path first
        return win_path_to_wsl_path(subprocess.check_output(["cygpath", "-w", win_path]).strip().decode())
    abspath_split = win_path.split(':/', maxsplit=1)
    # Starts with // to prevent MSYS2/Cygwin converting it to Windows path again 
    return '//mnt/' + abspath_split[0].lower() + '/' + abspath_split[1]

def handle_cmdline_arg(target_exec, arg):
    arg = arg.replace('\\', '/')    # Prevent argument to be escaped, and there is no \ in UNIX-styled arguments
    if arg.startswith('@'):         # Response file, but cctools do not support it.
        resp_file = arg[1:]
        if os.path.isfile(resp_file):
            with open(resp_file, 'r') as f:
                resp_file_args = shlex.split(f.read())
            retargs = []
            for temp_arg in resp_file_args:
                retargs += handle_cmdline_arg(target_exec, temp_arg)
            return retargs
        else:
            return [arg]
        
    if sys.platform not in ('win32', 'cygwin'):
        return [arg]
    if target_exec.endswith('-ld'):     # Linker
        if arg.startswith('-l') or arg.startswith('-L'):    # Link lib or link directory
            return [arg[:2] + win_path_to_wsl_path(arg[2:])]
    dir_name = os.path.dirname(arg)
    if dir_name and os.path.exists(dir_name):
        return [win_path_to_wsl_path(arg)]
    else:
        arg_split = arg.split('=', maxsplit=1)
        if len(arg_split) == 1:
            return [arg]
        else:
            dir_name = os.path.dirname(arg_split[1])
            if dir_name and os.path.exists(dir_name):
                return [arg_split[0] + '=' + win_path_to_wsl_path(arg_split[1])]
            else:
                return [arg]

def main(target, exec_name):
    target_exec = os.path.join(DIR, 'cctools-' + exec_name)
    run_args = []
    shell_eval_args = []
    if sys.platform in ('win32', 'cygwin'):
        run_args += ['wsl']
        wsl_distro = os.environ.get('LLVM_CROSS_WSL_DISTRO')
        if wsl_distro:
            run_args += ['-d', wsl_distro]
        # Pass remaining arguments via stdin becuase command line length is limited to only 32767 characters
        # Ref: https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessa
        run_args += ['--', 'sh', '-c', '$(cat -)']
        shell_eval_args += [win_path_to_wsl_path(target_exec)]
    for arg in sys.argv[1:]:
        shell_eval_args += handle_cmdline_arg(target_exec, arg)

    if sys.platform not in ('win32', 'cygwin'):
        os.execv(target_exec, [target_exec] + run_args + shell_eval_args)
    else:
        quote_args = [shlex.quote(arg) for arg in shell_eval_args]
        shell_eval_cmdline_bytes = (' '.join(quote_args)).encode('utf-8')
        exit(subprocess.run(run_args, input=shell_eval_cmdline_bytes).returncode)

File: wrapper/cctools_wrapper.go
package main

import (
	"io/ioutil"
	"os"
	"path/filepath"
	"runtime"
	"strings"

	shlex "github.com/kballard/go-shellquote"
)

func winPath2WSLPath(winPath string) string {
	if runtime.GOOS != "windows" {
		return winPath
	}
	winPath = strings.Replace(winPath, "\\", "/", -1)
	if !filepath.IsAbs(winPath) {
		return winPath
	}
	absPathSplit := strings.SplitN(winPath, ":/", 2)
	return "/mnt/" + strings.ToLower(absPathSplit[0]) + "/" + absPathSplit[1]
}

func ccToolsHandleCmdlineArg(targetExec string, arg string) []string {
	arg = strings.Replace(arg, "\\", "/", -1) // Prevent argument to be escaped, and there is no \ in UNIX-styled arguments
	if strings.HasPrefix(arg, "@") {          // Response file, but cctools do not support it.
		respFile := arg[1:]
		content, err := ioutil.ReadFile(respFile)
		if err == nil {
			tempArgs, err := shlex.Split(string(content))
			if err == nil {
				retArgs := []string{}
				for _, tempArg := range tempArgs {
					retArgs = append(retArgs, ccToolsHandleCmdlineArg(targetExec, tempArg)...)
				}
				return retArgs
			}
		}
		return []string{arg}
	}

	if runtime.GOOS != "windows" {
		return []string{arg}
	}
	if strings.HasSuffix(targetExec, "-ld") {
		// Linker
		if strings.HasPrefix(arg, "-l") || strings.HasPrefix(arg, "-L") {
			// Link lib or link directory
			return []string{arg[:2] + winPath2WSLPath(arg[2:])}
		}
	}
	dirName := filepath.Dir(arg)
	if len(dirName) > 0 && dirName != "." {
		_, err := os.Stat(dirName)
		if err == nil {
			return []string{winPath2WSLPath(arg)}
		}
	}
	argSplit := strings.SplitN(arg, "=", 2)
	if len(argSplit) == 2 {
		dirName = filepath.Dir(argSplit[1])
		_, err := os.Stat(dirName)
		if err == nil {
			return []string{argSplit[0] + "=" + winPath2WSLPath(argSplit[1])}
		}
	}
	return []string{arg}
}

func ccToolsWrapperMain(execDir string, target string, execName string, cmdArgv []string) {
	targetExec := filepath.Join(execDir, "cctools-"+execName)
	runArgs := []string{}
	shellEvelArgs := []string{}
	if runtime.GOOS == "windows" {
		wslDistro := os.Getenv("LLVM_CROSS_WSL_DISTRO")
		if len(wslDistro) > 0 {
			runArgs = append(runArgs, "-d", wslDistro)
		}
		// Pass remaining arguments via stdin becuase command line length is limited to only 32767 characters
		// Ref: https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessa
		runArgs = append(runArgs, "--", "sh", "-c", "$(cat -)")
		shellEvelArgs = append(shellEvelArgs, winPath2WSLPath(targetExec))
	}
	for _, arg := range cmdArgv[1:] {
		shellEvelArgs = append(shellEvelArgs, ccToolsHandleCmdlineArg(targetExec, arg)...)
	}
	if runtime.GOOS == "windows" {
		runCommand("wsl", runArgs, []byte(shlex.Join(shellEvelArgs...)))
	} else {
		runCommand(targetExec, append(runArgs, shellEvelArgs...), nil)
	}
}
